### PR TITLE
Forum channel

### PR DESCRIPTION
**Pull request by meister03**
- Helper Plugin - Send Message Check & createForumPost `Send_Message` permission Check
- Helpers - add Function createForumPost and move to new folder `forums` -> (more function in future will probably come)
- Transformers - add flags property
- Constants.Endpoints - add `FORUM_START`
- Enums - shared.ts - add `Channeltypes` 15 & 14

Missing:
- Any kind of test
- New Op & Statuscodes

### CODE REVIEW

File: helpers/channels/forums/createForumPost.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+export interface CreateForumMessage {
+  /** The message contents (up to 2000 characters) */
+  content?: string;
+  /** Embedded `rich` content (up to 6000 characters) */
+  embeds?: Embed[];
+  /** Allowed mentions for the message */
+  allowedMentions?: AllowedMentions;
+  /** Include to make your message a reply */
```
Comment (lts20050703) [inline helpers/channels/forums/createForumPost.ts:113]: Did you forgot to add a property or remove the comment?

File: helpers/channels/forums/createForumPost.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+              })),
+            };
+          }
+
+          return {
+            type: subComponent.type,
+            custom_id: subComponent.customId,
+            label: subComponent.label,
+            style: subComponent.style,
+            emoji: "emoji" in subComponent && subComponent.emoji
```
Comment (lts20050703) [inline helpers/channels/forums/createForumPost.ts:77]:
```suggestion
            emoji: subComponent.emoji
```

File: helpers/channels/forums/createForumPost.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+              ? {
+                id: subComponent.emoji.id?.toString(),
+                name: subComponent.emoji.name,
+                animated: subComponent.emoji.animated,
+              }
+              : undefined,
+            url: "url" in subComponent ? subComponent.url : undefined,
+            disabled: "disabled" in subComponent ? subComponent.disabled : undefined,
```
Comment (lts20050703) [inline helpers/channels/forums/createForumPost.ts:85]:
```suggestion
            url: subComponent.url ?? undefined,
            disabled: subComponent.disabled ?? undefined,
```

File: template/.gitignore [hunk 4]
```diff
@@ -3,3 +3,4 @@ fileloader.ts
 
 # Folder made by Kwik db to watch guild command versions
 db/
+node_modules
```
Comment (lts20050703) [inline template/.gitignore:6]: next time try not to accidentally include unrelated changes! with that being said, I see nothing wrong with this, it's just that it shouldn't be in a PR named "Forum channels"

**Review by lts20050703 — CHANGES_REQUESTED** [review]
Looks good to me! It's just looks like you forgot a property or a comment and a unrelated changes that shouldn't be in a PR named "Forum channel".

File: helpers/channels/forums/createForumPost.ts (hunk 2 above)
Comment (lts20050703) [reply helpers/channels/forums/createForumPost.ts:77]: note: I just want to know why "in", consider waiting for ITOH/Skill4Killz review before changing this

File: helpers/channels/forums/createForumPost.ts (hunk 3 above)
Comment (lts20050703) [reply helpers/channels/forums/createForumPost.ts:85]: note: I just want to know why "in", consider waiting for ITOH/Skill4Killz review before changing this

File: helpers/channels/forums/createForumPost.ts (hunk 1 above)
Comment (meister03) [reply helpers/channels/forums/createForumPost.ts:113]: Forgot to remove it

File: helpers/channels/forums/createForumPost.ts (hunk 2 above)
Comment (meister03) [reply helpers/channels/forums/createForumPost.ts:77]: Thats how it is done in the sendMessage.ts helper

Comment (splatterxl) [reply helpers/channels/forums/createForumPost.ts:77]: > note: I just want to know why "in"

probably to avoid setting it as undefined?

File: helpers/channels/forums/createForumPost.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+import { AllowedMentions, FileContent, MessageComponents } from "../../../types/mod.ts";
+import { Embed } from "../../../transformers/embed.ts";
+import { DiscordMessage } from "../../../types/discord.ts";
+import { MessageComponentTypes } from "../../../types/shared.ts";
+
+/** Creates a new public thread from an existing message. Returns a thread channel. */
+export async function createForumPost(
+  bot: Bot,
+  channelId: bigint,
+  options: createForumPostWithMessage,
```
Comment (Skillz4Killz) [inline helpers/channels/forums/createForumPost.ts:12]: We use PascalCase for interfaces, can u capitalize the `c` in create

File: helpers/channels/forums/createForumPost.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+      allowed_mentions: options.allowedMentions
+        ? {
+          parse: options.allowedMentions?.parse,
+          roles: options.allowedMentions?.roles?.map((id) => id.toString()),
+          users: options.allowedMentions?.users?.map((id) => id.toString()),
+          replied_user: options.allowedMentions?.repliedUser,
+        }
+        : undefined,
+      file: options.file,
+      components: options.components?.map((component) => ({
```
Comment (Skillz4Killz) [inline helpers/channels/forums/createForumPost.ts:34]: Why not use the transformers.reverse.component here?

File: helpers/channels/forums/createForumPost.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+          };
+        }),
+      })),
+    }
+  );
+
+  return bot.transformers.channel(bot, { channel: result, guildId: bot.transformers.snowflake(result.guild_id!) });
+}
+
+export interface createForumPostWithMessage extends CreateForumMessage {
```
Comment (Skillz4Killz) [inline helpers/channels/forums/createForumPost.ts:95]:
```suggestion
export interface CreateForumPostWithMessage extends CreateForumMessage {
```

**Review by Skillz4Killz — APPROVED** [review]

File: helpers/channels/forums/createForumPost.ts (hunk 5 above)
Comment (lts20050703) [inline helpers/channels/forums/createForumPost.ts:12]:
```suggestion
  options: CreateForumPostWithMessage,
```

File: helpers/channels/forums/createForumPost.ts (hunk 7 above)
Comment (lts20050703) [inline helpers/channels/forums/createForumPost.ts:95]:
```suggestion
export interface CreateForumPostWithMessage extends CreateForumMessage {
```

File: helpers/channels/forums/createForumPost.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                id: subComponent.emoji.id?.toString(),
+                name: subComponent.emoji.name,
+                animated: subComponent.emoji.animated,
+              }
+              : undefined,
+            url: "url" in subComponent ? subComponent.url : undefined,
+            disabled: "disabled" in subComponent ? subComponent.disabled : undefined,
+          };
+        }),
+      })),
```
Comment (lts20050703) [inline helpers/channels/forums/createForumPost.ts:88]:
```suggestion
      components: options.components?.map((component) => bot.transformers.reverse.component(component)),
```

**Review by lts20050703 — CHANGES_REQUESTED** [review]

**Comment by lts20050703** [comment]
Does this close #2161 ?

**Comment by lts20050703** [comment]
Does this one also close #2177?

File: helpers/channels/forums/createForumPost.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  channelId: bigint,
+  options: createForumPostWithMessage,
+) {
+  const result = await bot.rest.runMethod<DiscordChannel>(
+    bot.rest,
+    "post",
+    bot.constants.endpoints.FORUM_START(channelId),
+    {
+      name: options.name,
+      auto_archive_duration: options.autoArchiveDuration,
```
[the remark this answers is not in the document]
Comment (Skillz4Killz) [reply helpers/channels/forums/createForumPost.ts:18]: I still don't see the `reason` which in the options
